src/main/java/org/reaktivity/nukleus/socks/internal/stream/SocksServerFactory.java

```diff
@@ -0,0 +1,801 @@
+            int progress = offset;
+            if (handshakeRequest != null)
+            {
+                progress = handshakeRequest.limit();
+                onHandshakeRequest(handshakeRequest);
+            }
+            return progress;
+        }
+
+        private void onNetwork(
+            int msgTypeId,
+            DirectBuffer buffer,
+            int index,
+            int length)
+        {
+            switch (msgTypeId)
+            {
+                case BeginFW.TYPE_ID:
+                    final BeginFW begin = beginRO.wrap(buffer, index, index + length);
+                    onNetworkBegin(begin);
+                    break;
+                case DataFW.TYPE_ID:
+                    final DataFW data = dataRO.wrap(buffer, index, index + length);
+                    onNetworkData(data);
+                    break;
+                case EndFW.TYPE_ID:
+                    final EndFW end = endRO.wrap(buffer, index, index + length);
+                    onNetworkEnd(end);
+                    break;
+                case AbortFW.TYPE_ID:
+                    final AbortFW abort = abortRO.wrap(buffer, index, index + length);
+                    onNetworkAbort(abort);
+                    break;
+                case WindowFW.TYPE_ID:
+                    final WindowFW window = windowRO.wrap(buffer, index, index + length);
+                    onNetworkWindow(window);
+                    break;
+                case ResetFW.TYPE_ID:
+                    final ResetFW reset = resetRO.wrap(buffer, index, index + length);
+                    onNetworkReset(reset);
+                    break;
+            }
+        }
+
+        private void onNetworkData(
+            DataFW data)
+        {
+            final OctetsFW payload = data.payload();
+            initialBudget -= Math.max(data.length(), 0) + data.padding();
+            if (initialBudget < 0)
+            {
+                //TODO
+            }
+            else if (payload != null)
+            {
+                decodeTraceId = data.trace();
+                DirectBuffer buffer = payload.buffer();
+                int offset = payload.offset();
+                int limit = payload.limit();
+                this.authorization = data.authorization();
+                this.decodeTraceId = data.typeId();
+
+                if (bufferSlot != BufferPool.NO_SLOT)
+                {
+                    MutableDirectBuffer decodeBuffer = bufferPool.buffer(bufferSlot);
+                    decodeBuffer.putBytes(bufferSlotOffset, buffer, offset, limit - offset);
+                    bufferSlotOffset += limit - offset;
+                    buffer = decodeBuffer;
+                    offset = 0;
+                    limit = bufferSlot;
+                }
+
+                while(offset < limit)
+                {
+                    offset += decodeState.decode(buffer, offset, limit);
+                }
```

Comment:
This is good, but doesn't yet handle fragmentation.
You need to make sure that progress is being made or the decoder state changed.

Reply:
For the  `decoder` part, I changed the `decodeState`  inside the decode function. Do you mean that I need to create a `prevDecode` variable then compare it with the `decodeState` after the `decode` finish?

Reply:
Yes, see https://github.com/reaktivity/nukleus-http.java/blob/develop/src/main/java/org/reaktivity/nukleus/http/internal/stream/ServerAcceptStream.java#L434 for an example.